src/Common/src/TypeSystem/IL/Stubs/DelegateThunks.cs

```diff
@@ -141,6 +141,74 @@ public override string Name
     }
 
     /// <summary>
+    /// Invoke thunk for open delegates to instance methods. This kind of thunk
+    /// uses the first parameter as `this` that gets passed to the target instance method.
+    /// The thunk also performs virtual resolution if necessary.
+    /// This kind of delegates is typically created with Delegate.CreateDelegate
+    /// and MethodInfo.CreateDelegate at runtime.
+    /// </summary>
+    public sealed class DelegateInvokeOpenInstanceThunk : DelegateThunk
+    {
+        internal DelegateInvokeOpenInstanceThunk(DelegateInfo delegateInfo)
+            : base(delegateInfo)
+        {
+        }
+
+        public override MethodIL EmitIL()
+        {
+            var emitter = new ILEmitter();
+            ILCodeStream codeStream = emitter.NewCodeStream();
+
+            // Load all arguments except delegate's 'this'
+            TypeDesc boxThisType = null;
+            TypeDesc[] parameters = new TypeDesc[Signature.Length - 1];
+            for (int i = 0; i < Signature.Length; i++)
```

Comment:
This one is just more noticeable because of "new TypeDesc[Signature.Length - 1];" but I agree that all other conditions would need to be asserted too for completeness. I guess one way to avoid code duplication would be to add a static IsSupportedDelegateType(DelegateInfo) or something like this to the DelegateInvokeOpenInstanceThunk class. But that's not a big deal so feel free to ignore my comment.